### DISCUSSION TITLE
Auto-numbering elements and smart references to them.

### DISCUSSION THREAD

**Discussion by bwklein** [markdoc/markdoc#423] [Ideas]
Looking for a way to have things like Headings, Figures, Tables, etc. be automatically labeled and numbered.
So something like `== Heading One` would become "1. Heading One" as the heading text.

Then have an easy way to refer to that heading/section and have the link/reference show the proper reference to it.
As an example, in Asciidoc you would use something like `<<Heading One>>` to refer to that section and then in place of that shortcode you would see "Section 1" and that text would be linked back to an auto-generated anchor at the referenced section heading.

See an example of this functionality in AsciiDoc.

https://docs.asciidoctor.org/asciidoc/latest/macros/xref/#anchors  
and  
https://docs.asciidoctor.org/asciidoc/latest/macros/xref/#internal-cross-references

**Comment by mfix-stripe** [comment]
Hey @bwklein — you can auto label your headings and stuff by customizing the [heading node](https://markdoc.dev/docs/nodes)'s `transform` method. 

You could also automatically generate an ID for each heading ([as we have done here](https://github.com/markdoc/docs/blob/main/markdoc/nodes/heading.markdoc.js#L4)) and reference the heading that way.
